### PR TITLE
Flatten /prim and finish flatten

### LINKED ISSUE #937: Flatten scal / arr / mat
## Summary:
Flatten the directory structure by getting rid of the `scal`, `arr`, `mat` folders. The goal is to make the Math library easier to navigate for developers and users.

## Description:

Our current directory structure has looks like:
- `stan/math/`
- then `{prim, rev, fwd, mix}` indicating that we're dealing with no-autodiff, reverse mode, forward mode, or mixed mode
- then `{scal, arr, mat}` indicated that we're dealing with scalar, array, or matrix, all from the point of view of the Stan language. Scalar is just `double` or the autodiff variable, array includes `std::vector`, and matrix includes `Eigen`.
- then `err` for error checking, `fun` for functions, `meta` for template meta programs, `functor` for dealing with functors, `prob` for probability distributions
- then finally functions are in separate headers.

We'd like to get rid of the `{scal, arr, mat}` level. We had originally done this to support the use-case of a client of the Math library wanting to compile with only scalars, only scalars + the standard template library (`<vector>` for `std::vector`), or compile with all of that and Eigen. I swear we had users request this back when we had the stan-users list on google groups, but I could never find that thread again. The motivation was compile times and minimal includes. (This might have been back in the days when compiling under Windows was limited due to lack of a decent compiler; those days are over.) This has caused a fair amount of technical overhead keeping these things separate. It would really simplify our development if we can assume we have access to `std::vector` and `Eigen::Matrix` in every function. Right now, we include the template metaprograms in the very delicate, correct order to instantiate things the right way. So... the order of instantiation is different in `stan/math/prim/scal.hpp` vs `stan/math/prim/arr.hpp` vs `stan/math/prim/mat.hpp`. What we want to do is to instantiate the template metaprograms in the order that the `stan/math/*/mat.hpp` instantiates them in.

The nice thing is that there is a correct way to instantiate the template metaprograms and we know what that is. We need to trace through those dependencies.

### What the top-level includes will look like after we're done

```
1. stan/math/prim.hpp
2. stan/math/rev.hpp
3. stan/math/fwd.hpp
4. stan/math/mix.hpp
```

### The current top-level includes
```
1. stan/math/prim/arr.hpp
2. stan/math/prim/mat.hpp
3. stan/math/prim/scal.hpp
4. stan/math/rev/arr.hpp
5. stan/math/rev/mat.hpp
6. stan/math/rev/scal.hpp
7. stan/math/fwd/arr.hpp
8. stan/math/fwd/mat.hpp
9. stan/math/fwd/scal.hpp
10. stan/math/mix/arr.hpp
11. stan/math/mix/mat.hpp
12. stan/math/mix/scal.hpp
```

### Additional Information:
There are a few things we want to check:

- compile times don't go up for standard models
- everything still compiles

It's going to be a little tricky to do this. We should come up with some end-to-end tests for this to make sure it works on top of the lower-level unit tests that should continue to work.

Some unit tests will fail for the template metaprograms. The ones that will fail are for explicit checks that the base template metaprogram behaves in a certain way and we have specializations for `std::vector` or `Eigen::Matrix`. It's only once we include the specializations that the behavior will change.

The fix is to merge a bunch of files together. That's oversimplifying it, but that's essentially what we have to do. We just have to make sure things are merged in the right order. If done in that way, we'll maintain the correct behavior. I'll put together some examples below.

I believe we can script this so we don't have to do much by hand. In fact, I'm sure we can get a crude version working this way just by copying and pasting files in the right order.

There are going to be a lot of different ways we can do this. I'd like us to explore ways that keep the code-base working. Maybe even breaking this down into smaller chunks that we can merge as we go along. For example, one thing we may want to do is to merge all the template metaprograms into `mat` first and leave everything else alone.

We will need to modify `make/test-math-dependencies`. We will want to keep the rules for the autodiff types, but definitely not the ones for `scal`, `arr`, `mat`.

#### Current Version:
v2.17.0

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary

Final PR in the flatten of `arr, mat, scal`. This PR flattens the rest of `/prim` and removes the checks that checked for arr, scal, mat include order.

## Tests

/

## Checklist

- [x] Math issue Closes #937 

- [x] Copyright holder: Rok Češnovar

By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

**Comment by mcol** [comment]
What's the timeline for this? Is it waiting for the freeze first or do we need to push it before it?

**Comment by rok-cesnovar** [comment]
We havent really discussed it. The freeze will start on the 11th (not sure what time, we usually take midnight anywhere on earth- https://www.worldtimeserver.com/time-zones/aoe/ or smth like that). 

But I am not sure whether to merge this during the freeze or not. Its technically not a bugfix but its not a new feature either. It is a huge PR and once again the review will take some time. 

Personally I do want to get this in before the next release so that we release the library in a clean state. I dont wont to force it tho.

**Comment by mcol** [comment]
I would also like to see it make it, but given the size and impact on other PRs it would be more convenient to know that it could go in just after the freeze. @seantalts would that be possible?

In any case, I'm starting reviewing it now, let's see how far we can get.

**Comment by rok-cesnovar** [comment]
Thanks Marco!

File: stan/math/prim/fun/mdivide_right_ldlt.hpp [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
 
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/LDLT_factor.hpp>
-#include <stan/math/prim/mat/fun/mdivide_left_ldlt.hpp>
-#include <stan/math/prim/mat/fun/transpose.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/LDLT_factor.hpp>
+#include <stan/math/prim/fun/mdivide_left_ldlt.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
 #include <stan/math/prim/err.hpp>
```
Comment (mcol) [inline stan/math/prim/fun/mdivide_right_ldlt.hpp]: Move the err header first.

File: stan/math/prim/fun/mdivide_right_spd.hpp [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
+#ifndef STAN_MATH_PRIM_FUN_MDIVIDE_RIGHT_SPD_HPP
+#define STAN_MATH_PRIM_FUN_MDIVIDE_RIGHT_SPD_HPP
 
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/mdivide_left_spd.hpp>
-#include <stan/math/prim/mat/fun/transpose.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/mdivide_left_spd.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
 #include <stan/math/prim/err.hpp>
```
Comment (mcol) [inline stan/math/prim/fun/mdivide_right_spd.hpp]: err header first.

File: stan/math/prim/fun/mdivide_right_tri.hpp [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
 
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/mdivide_left_tri.hpp>
-#include <stan/math/prim/mat/fun/promote_common.hpp>
-#include <stan/math/prim/mat/fun/transpose.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/mdivide_left_tri.hpp>
+#include <stan/math/prim/fun/promote_common.hpp>
+#include <stan/math/prim/fun/transpose.hpp>
 #include <stan/math/prim/err.hpp>
```
Comment (mcol) [inline stan/math/prim/fun/mdivide_right_tri.hpp]: err header first.

File: stan/math/prim/fun/erfc.hpp [hunk 4]
```diff
@@ -0,0 +1,62 @@
+#ifndef STAN_MATH_PRIM_FUN_ERFC_HPP
+#define STAN_MATH_PRIM_FUN_ERFC_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/fun/erfc.hpp>
```
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply stan/math/prim/fun/erfc.hpp]: Thanks!

**Comment by rok-cesnovar** [comment]
I addressed all comments in the first batch. 

I would definitely wait for at least https://github.com/stan-dev/math/pull/1558 to get merged before merging this one. The rest are currently either not ready yet or I would consider as bugfix PRs.

File: stan/math/prim/prob/wiener_lpdf.hpp [hunk 5]
```diff
@@ -31,11 +31,11 @@
...
 #define STAN_MATH_PRIM_PROB_WIENER_LPDF_HPP
 
-#include <stan/math/prim/scal/fun/constants.hpp>
-#include <stan/math/prim/scal/fun/square.hpp>
-#include <stan/math/prim/scal/fun/value_of.hpp>
+#include <stan/math/prim/fun/constants.hpp>
+#include <stan/math/prim/fun/square.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
 #include <stan/math/prim/err.hpp>
-#include <stan/math/prim/scal/fun/size_zero.hpp>
```
Comment (mcol) [inline stan/math/prim/prob/wiener_lpdf.hpp]: Put err.hpp header first.

File: stan/math/rev/core/operator_unary_plus.hpp [hunk 6]
```diff
@@ -2,9 +2,9 @@
 #define STAN_MATH_REV_CORE_OPERATOR_UNARY_PLUS_HPP
 
 #include <stan/math/rev/core/var.hpp>
-#include <stan/math/prim/scal/fun/is_nan.hpp>
+#include <stan/math/prim/fun/is_nan.hpp>
 #include <stan/math/rev/core/precomp_v_vari.hpp>
-#include <stan/math/prim/scal/fun/constants.hpp>
+#include <stan/math/prim/fun/constants.hpp>
```
Comment (mcol) [inline stan/math/rev/core/operator_unary_plus.hpp]: prim headers together.

File: stan/math/rev/fun/Eigen_NumTraits.hpp [hunk 7]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_EIGEN_NUMTRAITS_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/core/std_numeric_limits.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/Eigen_NumTraits.hpp]: Put the rev headers together.

**Comment by bob-carpenter** [comment]
#1525 is another big one waiting to be merged but that's probably not going to be ready to go for next PR (isn't code freeze today the 11th?).

File: stan/math/rev/fun/acosh.hpp [hunk 8]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_ACOSH_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/acosh.hpp>
+#include <stan/math/prim/fun/acosh.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/acosh.hpp]: rev headers together.

File: stan/math/rev/fun/asinh.hpp [hunk 9]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_ASINH_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/asinh.hpp>
+#include <stan/math/prim/fun/asinh.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/asinh.hpp]: rev headers together.

File: stan/math/rev/fun/atanh.hpp [hunk 10]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_ATANH_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/atanh.hpp>
+#include <stan/math/prim/fun/atanh.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/atanh.hpp]: rev headers together.

File: stan/math/rev/fun/cbrt.hpp [hunk 11]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_CBRT_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/cbrt.hpp>
+#include <stan/math/prim/fun/cbrt.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/cbrt.hpp]: rev headers together.

File: stan/math/rev/fun/determinant.hpp [hunk 12]
```diff
@@ -3,7 +3,7 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/prim/err.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/determinant.hpp]: rev headers together.

File: stan/math/rev/fun/erfc.hpp [hunk 13]
```diff
@@ -2,8 +2,8 @@
 #define STAN_MATH_REV_FUN_ERFC_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/constants.hpp>
-#include <stan/math/prim/scal/fun/erfc.hpp>
+#include <stan/math/prim/fun/constants.hpp>
+#include <stan/math/prim/fun/erfc.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/erfc.hpp]: rev headers together.

File: stan/math/rev/fun/expm1.hpp [hunk 14]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_EXPM1_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/expm1.hpp>
+#include <stan/math/prim/fun/expm1.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/expm1.hpp]: rev headers together.

File: stan/math/rev/fun/fdim.hpp [hunk 15]
```diff
@@ -2,8 +2,8 @@
 #define STAN_MATH_REV_FUN_FDIM_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/constants.hpp>
-#include <stan/math/prim/scal/fun/is_any_nan.hpp>
+#include <stan/math/prim/fun/constants.hpp>
+#include <stan/math/prim/fun/is_any_nan.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/fdim.hpp]: rev headers together.

File: stan/math/rev/fun/fmin.hpp [hunk 16]
```diff
@@ -3,9 +3,9 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/rev/core.hpp>
-#include <stan/math/prim/scal/fun/constants.hpp>
+#include <stan/math/prim/fun/constants.hpp>
 #include <stan/math/rev/fun/is_nan.hpp>
-#include <stan/math/prim/scal/fun/is_nan.hpp>
+#include <stan/math/prim/fun/is_nan.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/fmin.hpp]: rev headers together.

File: stan/math/rev/fun/grad.hpp [hunk 17]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_GRAD_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/fun/Eigen_NumTraits.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/grad.hpp]: rev headers together.

File: stan/math/rev/fun/grad_inc_beta.hpp [hunk 18]
```diff
@@ -3,9 +3,9 @@
...
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/rev/core.hpp>
-#include <stan/math/prim/scal/fun/beta.hpp>
-#include <stan/math/prim/scal/fun/grad_2F1.hpp>
-#include <stan/math/prim/scal/fun/value_of.hpp>
+#include <stan/math/prim/fun/beta.hpp>
+#include <stan/math/prim/fun/grad_2F1.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
 #include <stan/math/rev/fun/exp.hpp>
 #include <stan/math/rev/fun/fabs.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/grad_inc_beta.hpp]: rev headers together.

File: stan/math/rev/fun/inc_beta.hpp [hunk 19]
```diff
@@ -3,10 +3,10 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/rev/core.hpp>
-#include <stan/math/prim/scal/fun/grad_reg_inc_beta.hpp>
+#include <stan/math/prim/fun/grad_reg_inc_beta.hpp>
 #include <stan/math/rev/fun/pow.hpp>
-#include <stan/math/prim/scal/fun/beta.hpp>
-#include <stan/math/prim/scal/fun/digamma.hpp>
+#include <stan/math/prim/fun/beta.hpp>
+#include <stan/math/prim/fun/digamma.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/inc_beta.hpp]: rev headers together.

File: stan/math/rev/fun/inverse.hpp [hunk 20]
```diff
@@ -1,12 +1,12 @@
...
 #define STAN_MATH_REV_FUN_INVERSE_HPP
 
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/prim/err.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/inverse.hpp]: rev headers together.

File: stan/math/rev/fun/log1m_inv_logit.hpp [hunk 21]
```diff
@@ -2,8 +2,8 @@
 #define STAN_MATH_REV_FUN_LOG1M_INV_LOGIT_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/log1m_inv_logit.hpp>
-#include <stan/math/prim/scal/fun/inv_logit.hpp>
+#include <stan/math/prim/fun/log1m_inv_logit.hpp>
+#include <stan/math/prim/fun/inv_logit.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/core/precomp_v_vari.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/log1m_inv_logit.hpp]: rev headers together.

File: stan/math/rev/fun/log1p_exp.hpp [hunk 22]
```diff
@@ -3,7 +3,7 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/rev/core.hpp>
-#include <stan/math/prim/scal/fun/log1p_exp.hpp>
+#include <stan/math/prim/fun/log1p_exp.hpp>
 #include <stan/math/rev/fun/calculate_chain.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/log1p_exp.hpp]: rev headers together.

File: stan/math/rev/fun/log_determinant.hpp [hunk 23]
```diff
@@ -3,8 +3,8 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/prim/err.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/log_determinant.hpp]: rev headers together.

File: stan/math/rev/fun/log_inv_logit.hpp [hunk 24]
```diff
@@ -2,8 +2,8 @@
 #define STAN_MATH_REV_FUN_LOG_INV_LOGIT_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/inv_logit.hpp>
-#include <stan/math/prim/scal/fun/log_inv_logit.hpp>
+#include <stan/math/prim/fun/inv_logit.hpp>
+#include <stan/math/prim/fun/log_inv_logit.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/core/precomp_v_vari.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/log_inv_logit.hpp]: rev headers together.

File: stan/math/rev/fun/log_mix.hpp [hunk 25]
```diff
@@ -2,8 +2,8 @@
 #define STAN_MATH_REV_FUN_LOG_MIX_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/value_of.hpp>
-#include <stan/math/prim/scal/fun/log_mix.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <stan/math/prim/fun/log_mix.hpp>
 #include <stan/math/rev/fun/value_of.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/log_mix.hpp]: rev headers together.

File: stan/math/rev/fun/log_rising_factorial.hpp [hunk 26]
```diff
@@ -3,8 +3,8 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/rev/core.hpp>
-#include <stan/math/prim/scal/fun/log_rising_factorial.hpp>
-#include <stan/math/prim/scal/fun/digamma.hpp>
+#include <stan/math/prim/fun/log_rising_factorial.hpp>
+#include <stan/math/prim/fun/digamma.hpp>
 #include <stan/math/rev/fun/digamma.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/log_rising_factorial.hpp]: rev headers together.

File: stan/math/rev/fun/log_softmax.hpp [hunk 27]
```diff
@@ -2,11 +2,11 @@
...
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/log_softmax.hpp>
-#include <stan/math/prim/mat/fun/softmax.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/log_softmax.hpp>
+#include <stan/math/prim/fun/softmax.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/log_softmax.hpp]: rev headers together.

File: stan/math/rev/fun/logit.hpp [hunk 28]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_LOGIT_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/logit.hpp>
+#include <stan/math/prim/fun/logit.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/core/precomp_v_vari.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/logit.hpp]: rev headers together.

File: stan/math/rev/fun/matrix_power.hpp [hunk 29]
```diff
@@ -5,8 +5,8 @@
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/fun/value_of.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/matrix_power.hpp]: rev headers together.

File: stan/math/rev/fun/mdivide_left.hpp [hunk 30]
```diff
@@ -3,8 +3,8 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/prim/err.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/mdivide_left.hpp]: rev headers together.

File: stan/math/rev/fun/multiply_lower_tri_self_transpose.hpp [hunk 31]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_MULTIPLY_LOWER_TRI_SELF_TRANSPOSE_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/multiply_lower_tri_self_transpose.hpp]: rev headers together.

File: stan/math/rev/fun/ordered_constrain.hpp [hunk 32]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_ORDERED_CONSTRAIN_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/functor/adj_jac_apply.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/ordered_constrain.hpp]: rev headers together.

File: stan/math/rev/fun/positive_ordered_constrain.hpp [hunk 33]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_POSITIVE_ORDERED_CONSTRAIN_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/functor/adj_jac_apply.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/positive_ordered_constrain.hpp]: rev headers together.

File: stan/math/rev/fun/quad_form.hpp [hunk 34]
```diff
@@ -3,10 +3,10 @@
...
 #include <stan/math/rev/core.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
-#include <stan/math/prim/mat/fun/value_of.hpp>
-#include <stan/math/prim/mat/fun/quad_form.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <stan/math/prim/fun/quad_form.hpp>
 #include <stan/math/prim/err.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/quad_form.hpp]: Move prim/err.hpp up.

File: stan/math/rev/fun/rows_dot_product.hpp [hunk 35]
```diff
@@ -3,7 +3,7 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/prim/err.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/rows_dot_product.hpp]: rev headers together.

File: stan/math/rev/fun/simplex_constrain.hpp [hunk 36]
```diff
@@ -2,8 +2,8 @@
 #define STAN_MATH_REV_FUN_SIMPLEX_CONSTRAIN_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/scal/fun/inv_logit.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/inv_logit.hpp>
 #include <stan/math/rev/functor/adj_jac_apply.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/simplex_constrain.hpp]: rev headers together.

File: stan/math/rev/fun/softmax.hpp [hunk 37]
```diff
@@ -2,9 +2,9 @@
 #define STAN_MATH_REV_FUN_SOFTMAX_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
-#include <stan/math/prim/mat/fun/softmax.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <stan/math/prim/fun/softmax.hpp>
 #include <stan/math/rev/functor/adj_jac_apply.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/softmax.hpp]: rev headers together.

File: stan/math/rev/fun/tgamma.hpp [hunk 38]
```diff
@@ -2,8 +2,8 @@
 #define STAN_MATH_REV_FUN_TGAMMA_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/digamma.hpp>
-#include <stan/math/prim/scal/fun/tgamma.hpp>
+#include <stan/math/prim/fun/digamma.hpp>
+#include <stan/math/prim/fun/tgamma.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/tgamma.hpp]: rev headers together.

File: stan/math/rev/fun/value_of_rec.hpp [hunk 39]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUN_VALUE_OF_REC_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/scal/fun/value_of_rec.hpp>
+#include <stan/math/prim/fun/value_of_rec.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/fun/value_of_rec.hpp]: rev headers together.

File: stan/math/rev/functor/coupled_ode_system.hpp [hunk 40]
```diff
@@ -3,7 +3,7 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/prim/functor/coupled_ode_system.hpp>
-#include <stan/math/prim/arr/fun/value_of.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
 #include <stan/math/prim/err.hpp>
 #include <stan/math/rev/fun/value_of_rec.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/functor/coupled_ode_system.hpp]: rev headers together.

File: stan/math/rev/functor/gradient.hpp [hunk 41]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUNCTOR_GRADIENT_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/functor/gradient.hpp]: rev headers together.

File: stan/math/rev/functor/jacobian.hpp [hunk 42]
```diff
@@ -2,7 +2,7 @@
 #define STAN_MATH_REV_FUNCTOR_JACOBIAN_HPP
 
 #include <stan/math/rev/meta.hpp>
-#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
 #include <stan/math/rev/core.hpp>
```
Comment (mcol) [inline stan/math/rev/functor/jacobian.hpp]: rev headers together.

File: stan/math/rev/functor/map_rect_reduce.hpp [hunk 43]
```diff
@@ -3,7 +3,7 @@
 
 #include <stan/math/rev/meta.hpp>
 #include <stan/math/prim/functor/map_rect_reduce.hpp>
-#include <stan/math/prim/mat/fun/typedefs.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
 #include <stan/math/rev/core/var.hpp>
 #include <stan/math/rev/fun/typedefs.hpp>
```
Comment (mcol) [inline stan/math/rev/functor/map_rect_reduce.hpp]: rev headers together.

**Comment by rok-cesnovar** [comment]
> #1525 is another big one waiting to be merged but that's probably not going to be ready to go for next 

I was assuming it wont be ready, but yeah I am happ to wait for that PR too, if needed.

> (isn't code freeze today the 11th?).

Yes, 11th anywhere on the earth. If this gets approved today I will merge it once the deadline passes.

**Comment by seantalts** [comment]
Agreed, thanks Rok! and thanks @mcol for stepping up to review :)

**Comment by rok-cesnovar** [comment]
Thanks Marco for all the reviews. You really made this process as smooth as it could have been.

**Pull request merged**